File: frontend/src/types/listing.ts
export interface Listing {
  id: string;
  title: string;
  description: string;
  category: string;
  location_city: string;
  location_country: string;
  price_per_night: number;
  max_guests: number;
  bedrooms: number;
  bathrooms: number;
  /** Comma-separated string from the backend */
  amenities: string;
  image_url: string;
  /** Newline-separated extra image URLs */
  extra_images: string;
  host_id: string;
  created_at: string;
}

export type ListingCategory =
  | "Beach"
  | "Mountains"
  | "Countryside"
  | "Pools"
  | "Islands"
  | "Lake"
  | "Skiing"
  | "Castles"
  | "Caves"
  | "Camping"
  | "Arctic"
  | "Desert"
  | "Barns"
  | "Lux";

export interface ListingFilters {
  category?: string;
  location?: string;
  min_price?: number;
  max_price?: number;
  guests?: number;
}

File: backend/app/schemas/listing.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class ListingCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    description: str = Field(..., min_length=1)
    category: str = Field(..., min_length=1, max_length=50)
    location_city: str = Field(..., min_length=1, max_length=100)
    location_country: str = Field(..., min_length=1, max_length=100)
    price_per_night: float = Field(..., gt=0)
    max_guests: int = Field(..., gt=0)
    bedrooms: int = Field(default=1, ge=1)
    bathrooms: int = Field(default=1, ge=1)
    amenities: str = Field(default="")
    image_url: str = Field(..., min_length=1)
    extra_images: str = Field(default="")


class ListingUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None, min_length=1)
    category: Optional[str] = Field(None, min_length=1, max_length=50)
    location_city: Optional[str] = Field(None, min_length=1, max_length=100)
    location_country: Optional[str] = Field(None, min_length=1, max_length=100)
    price_per_night: Optional[float] = Field(None, gt=0)
    max_guests: Optional[int] = Field(None, gt=0)
    bedrooms: Optional[int] = Field(None, ge=1)
    bathrooms: Optional[int] = Field(None, ge=1)
    amenities: Optional[str] = None
    image_url: Optional[str] = Field(None, min_length=1)
    extra_images: Optional[str] = None


class ListingResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    title: str
    description: str
    category: str
    location_city: str
    location_country: str
    price_per_night: float
    max_guests: int
    bedrooms: int
    bathrooms: int
    amenities: str  # comma-separated, parsed by frontend
    image_url: str
    extra_images: str  # newline-separated, parsed by frontend
    host_id: str
    created_at: datetime

